### PR TITLE
I2C major change: Refactoring, Rewritten Unit Tests etc

### PR DESCRIPTION
**Pull request by RichardKroesen**
**Warning**: There are a lot of structural I2C changes, but also names and sensor testcases are changed.

### Why this PR?
For I2C we had for a long time struggles to make it modular to make it available for sensor modules communication and for mainboard-hub internal system communication. Besides there we are a lot challenges because I2C has a strong dependency in the library. Lastly for testing purposes creating mocks for different I2C implementation abstraction layers, we needed loose coupling dependency injection.

 ### What has changed:
- I2C_driver class: is a wrapper for TwoWire. 
- I2C_sensor_driver: is an inherent class of I2C_driver for higher level functions implementations to make it useable for sensor communication.
- Mock_I2C_peripheral: for testing the I2C_sensor_driver
- Mock_I2C_sensor_driver created for creating sensor test cases.
- All the sensor testcases.
- hal i2c class diagram, checkout if you are curious for visual representation of current I2C relations. 

### Arguments for certain decisions
- Virtual functions inside I2C_sensor_driver class: makes the usage of mocking really practical, and if for some reason the implementation a slight different implementation there is a workaround for it. If somebody has a better way or suggestion, I'm more than open for it.
- Why I2C_driver as a wrapper? Having a wrapper makes testing and building on top more convenient. With this class we have atleast one constant dependency which does not change whether you are using it for testing or for end product.

### Pay Attention to:
- Make sure to verify that testcases of sensor modules are properly modified. I checked them aswell, but could overlook or missed cases during the change. 
- hal_i2c structure idea. 
- Any suggestions or feedback is more than welcome.

### The future of this functionality:
- Implementation of the main-hub communication using the I2C_driver class.
- Adding more abstraction layers using I2C_driver class for modules, like breathing module or actuator hubs.

### CODE REVIEW

File: modules/hal_gpio/src/gpio.cpp [hunk 1]
```diff
@@ -30,67 +30,68 @@
```
Comment (Hoog-V) [inline modules/hal_gpio/src/gpio.cpp:30]: This isn't refactored according to the google style guide guidelines... There shouldn't be indents after namespace

File: modules/hal_gpio/src/gpio.cpp [hunk 2]
```diff
@@ -30,67 +30,68 @@
...
+    }
 
-uint8_t GetGPIOPinFunction(GPIOPort port_num, uint8_t pin_num) {
-  if (pin_num % 2 == 0) {
-    return PORT->Group[port_num].PMUX[pin_num >> 1].bit.PMUXE;
-  } else {
-    return PORT->Group[port_num].PMUX[pin_num >> 1].bit.PMUXO;
-  }
-}
 }  // namespace hal::gpio
```
Comment (Hoog-V) [inline modules/hal_gpio/src/gpio.cpp:97]: According the google style guide an empty line should be added to end of the file

**Review by Hoog-V — CHANGES_REQUESTED** [review]
Had a quick look without running on my own pc.. these are the things that need to be changed:
- Github workflow change should be removed
- You need to check your google style guide with cpplint command (see workflow for exact command). Because there are some things which are not refactored according to google style guide

You're gonna get more feedback but these are the quick points I found...

File: .vscode/settings.json [hunk 3]
```diff
@@ -0,0 +1,21 @@
+{
```
[the remark this answers is not in the document]
Comment (RichardKroesen) [reply .vscode/settings.json:1]: Jup, overlooked.

**Comment by Hoog-V** [comment]
I  think this could be a lot simpler.. This doesn't compile will never compile.. Well maybe, but it isn't worth it Why? There are things which are wrapped in classes which shouldn't be in classes.. It is not java :) 

For example: Callback functions, like the functions passed in to the onreceive or onrequest functions should be separate functions outside of class, not in the class itself.. If you want to protect the function from outside access (there is a pretty word for this... I don't remember, I believe it was something like namespace pollution) , put it in to the source file instead of the header file. 

Another thing that doesn't feel right.. is the fact that things are wrapped multiple times. Why wrap a present multiple times in wrapping paper. Isn't it wasteful of the wrapping paper? Instead I would suggest the following:

The class diagram could be a lot flatter by just accepting the I2C_Driver as wrapper. Which already has all the functions we need. Why do you make three specific classes with specific api for hub, slave and what not. It complicates things, which is the thing we want to prevent. Because complexity can become a burden. I already have difficulty with understanding what you wrote, can you imagine what a new project group will think of our work.. 

Every complexity we add the more difficult it will become to trace back bugs during the integration phase of the project :) and also the more chance the next group will abandon our work... 

TLDR;

- Keep the I2C_Driver wrapper
- Try a functional programming approach for your i2c slave implementation
(if you write your i2c slave right it could fit in 1 or maybe 2 source files :) )
- Don't wrap registers in classes. Use structs instead, as they don't need to have any behavior :)
- KISS principle, loose coupling is IMO less important than KISS (This is an opinion!)

These are my points for now... They might be a bit controversial.. I'm open for discussion. Though most of the points mentioned are from a quick look at your code when loaded in to SensorHub_FW example with small changes trying to make it work...

**Comment by RichardKroesen** [comment]
> I think this could be a lot simpler.. This doesn't compile will never compile.. Well maybe, but it isn't worth it Why? There are things which are wrapped in classes which shouldn't be in classes.. It is not java :)
> 
> For example: Callback functions, like the functions passed in to the onreceive or onrequest functions should be separate functions outside of class, not in the class itself.. If you want to protect the function from outside access (there is a pretty word for this... I don't remember, I believe it was something like namespace pollution) , put it in to the source file instead of the header file.
> 
> Another thing that doesn't feel right.. is the fact that things are wrapped multiple times. Why wrap a present multiple times in wrapping paper. Isn't it wasteful of the wrapping paper? Instead I would suggest the following:
> 
> The class diagram could be a lot flatter by just accepting the I2C_Driver as wrapper. Which already has all the functions we need. Why do you make three specific classes with specific api for hub, slave and what not. It complicates things, which is the thing we want to prevent. Because complexity can become a burden. I already have difficulty with understanding what you wrote, can you imagine what a new project group will think of our work..
> 
> Every complexity we add the more difficult it will become to trace back bugs during the integration phase of the project :) and also the more chance the next group will abandon our work...
> 
> TLDR;
> 
> * Keep the I2C_Driver wrapper
> * Try a functional programming approach for your i2c slave implementation
>   (if you write your i2c slave right it could fit in 1 or maybe 2 source files :) )
> * Don't wrap registers in classes. Use structs instead, as they don't need to have any behavior :)
> * KISS principle, loose coupling is IMO less important than KISS (This is an opinion!)
> 
> These are my points for now... They might be a bit controversial.. I'm open for discussion. Though most of the points mentioned are from a quick look at your code when loaded in to SensorHub_FW example with small changes trying to make it work...

Thank you, Victor, for your honest feedback.
_(Excuses for being extensiveness, in this cause it's important to share a complete explanation)._

First things first, I can build this whole project and run all tests (I check this every time before making a commit).

Secondly, complexity is relative, and I think it only looks that way to you right now because I pushed everything at once. INP needs to and will change; it's just a rough concept at this point.

The purpose of wrapping is to get a stable underlying basis (low-level driver) before moving to the next implementation in the hierarchy of abstraction layers. The thing is that the macro for changing the underlying implementation of the I2C peripheral (mock or TwoWire object) creates a really painful situation if you try to build on top of it. Maybe you didn't encounter this, but I got nasty linking errors. Well, those are harder to debug than compiled code.

As I wrote before, this is my argument for choosing a wrapper around the I2C_PERIPHERAL_T macro. I also tried other solutions that were relatively more complex, like interfacing and templates.

By the way, the idea about callback functions is a good one. Thank you for suggesting it.

About INP, I stay with this class implementation because we indeed don't want unnecessary complexity. Separating the hub and master classes can reduce complexity, improve modularity and comply with SOLID principles. Functionaly, they are different, but they have the same peripheral driver. 

The 'extra' layers are not mindless or for the sake of it. They are meant as an abstraction of details to counteract complexity.

As a side note, in previous discussions about the i2c_hal, we talked about separating the master and hub on the low-level driver. I still wanted a seperation but on the higher level, because they still need the same low-level driver.

**To sum up:**
This is my response to share my point of view. 
_(Excuses for being extensiveness, in this cause it's important to share a complete explanation)._

- I will analyze where the CMake build trouble comes from.
- I will improve INP further.
- We better discuss this further in person.
